pkg/providers/imagefamily/image.go

```diff
@@ -91,57 +86,116 @@ func (p *Provider) KubeServerVersion(ctx context.Context) (string, error) {
-		for pager.More() {
-			page, err := pager.NextPage(context.Background())
-			if err != nil {
-				return "", err
-			}
-			for _, imageVersion := range page.CommunityGalleryImageVersionList.Value {
-				if lo.IsEmpty(topImageVersionCandidate) || imageVersion.Properties.PublishedDate.After(*topImageVersionCandidate.Properties.PublishedDate) {
-					topImageVersionCandidate = *imageVersion
-				}
-			}
+// Get returns Image ID for the given instance type. Images may vary due to architecture, accelerator, etc
+func (p *Provider) Get(ctx context.Context, nodeClass *v1alpha2.AKSNodeClass, instanceType *cloudprovider.InstanceType, imageFamily ImageFamily) (string, error) {
+	defaultImages := imageFamily.DefaultImages()
+	for _, defaultImage := range defaultImages {
+		if err := instanceType.Requirements.Compatible(defaultImage.Requirements, v1alpha2.AllowUndefinedLabels); err == nil {
+			return p.GetLatestImageID(ctx, defaultImage, nodeClass.Spec.GetImageVersion())
 		}
-		versionName = lo.FromPtr(topImageVersionCandidate.Name)
 	}
 
-	selectedImageID := BuildImageID(publicGalleryURL, communityImageName, versionName)
-	if p.cm.HasChanged(key, selectedImageID) {
-		logging.FromContext(ctx).With("image-id", selectedImageID).Info("discovered new image id")
+	return "", fmt.Errorf("no compatible images found for instance type %s", instanceType.Name)
+}
+
+// GetLatestImageID returns the latest image ID for the given image version and image specification
+func (p *Provider) GetLatestImageID(ctx context.Context, defaultImage DefaultImageOutput, imageVersion string) (string, error) {
+	// Managed Karpenter will use the AKS Managed Shared Image Galleries
+	if options.FromContext(ctx).ManagedKarpenter {
+		return p.getImageIDSIG(ctx, defaultImage, imageVersion)
 	}
-	p.imageCache.Set(key, selectedImageID, imageExpirationInterval)
-	return selectedImageID, nil
+	// Self Hosted Karpenter will use the Community Image Galleries, which are public and have lower scaling limits
+	return p.getImageIDCIG(ctx, defaultImage.PublicGalleryURL, defaultImage.ImageDefinition, imageVersion)
 }
 
-func BuildImageID(publicGalleryURL, communityImageName, imageVersion string) string {
-	return fmt.Sprintf(imageIDFormat, publicGalleryURL, communityImageName, imageVersion)
+// getImageIDSIG retrieves the shared image gallery image URL in the format:
+// /subscriptions/{subscriptionID}/resourceGroups/{resourceGroup}/providers/Microsoft.Compute/galleries/{galleryName}/images/{imageDefinition}/versions/{imageVersion}.
+// If imageVersion is an empty string, it fetches the latest version of the image and caches it for imageExpirationInterval (3 days).
+func (p *Provider) getImageIDSIG(ctx context.Context, imgStub DefaultImageOutput, imageVersion string) (string, error) {
+	key := fmt.Sprintf(sharedImageGalleryImageIDFormat, options.FromContext(ctx).SIGSubscriptionID, imgStub.GalleryResourceGroup, imgStub.GalleryName, imgStub.ImageDefinition, imageVersion)
```

Comment:
Why does subscription ID need to be part of the key?

Reply:
We could reconstruct the full id without the cache including subscription id. Especially since sub is a constant. Good catch.